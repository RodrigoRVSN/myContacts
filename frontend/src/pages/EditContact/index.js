import React from 'react';
import { ContactForm } from '../../components/ContactForm';
import { PageHeader } from '../../components/PageHeader';

export function EditContact() {
  return (
    <>
      <PageHeader title="Editar Contato" />
      <ContactForm buttonLabel="Salvar" />
    </>
  );
}
