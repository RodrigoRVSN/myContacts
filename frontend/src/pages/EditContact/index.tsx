import React from 'react';
import { ContactForm } from '../../components/ContactForm';
import { Loader } from '../../components/Loader';
import { PageHeader } from '../../components/PageHeader';
import { useEditContact } from './useEditContact';

export default function EditContact() {
  const {
    isLoading, contactName, contactForm, handleSubmit,
  } = useEditContact();

  return (
    <>
      <Loader isLoading={isLoading} />

      <PageHeader title={isLoading ? 'Carregando...' : `Editar ${contactName}`} />

      <ContactForm buttonLabel="Salvar" ref={contactForm} onSubmit={handleSubmit} />
    </>
  );
}
