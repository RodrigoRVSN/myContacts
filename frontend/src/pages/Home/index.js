/* eslint-disable no-nested-ternary */
import React, {
  useCallback, useEffect, useMemo, useState,
} from 'react';
import { Link } from 'react-router-dom';
import {
  Container,
  Header,
  Card,
  InputSearchContainer,
  ListHeader,
  ErrorContainer,
  NoContactsContainer,
  NotFoundFilteredContacts,
} from './styles';

import arrow from '../../assets/images/icons/arrow.svg';
import edit from '../../assets/images/icons/edit.svg';
import trash from '../../assets/images/icons/trash.svg';
import sad from '../../assets/images/sad.svg';
import emptyBox from '../../assets/images/empty-box.svg';
import magnifierQuestion from '../../assets/images/magnifier-question.svg';

import { Loader } from '../../components/Loader';
import ContactsService from '../../services/ContactsService';
import Button from '../../components/Button';

export function Home() {
  const [contacts, setContacts] = useState([]);
  const [orderBy, setOrderBy] = useState('asc');
  const [searchTerm, setSearchTerm] = useState('');
  const [loading, setLoading] = useState(true);
  const [hasError, setError] = useState(false);

  const filteredContacts = useMemo(
    () =>
      contacts.filter((contact) =>
        contact.name.toLowerCase().includes(searchTerm.toLowerCase())),
    [contacts, searchTerm],
  );

  const loadContacts = useCallback(async () => {
    try {
      setLoading(true);
      const contactsList = await ContactsService.listContacts(orderBy);

      setError(false);
      setContacts(contactsList);
    } catch (error) {
      setError(true);
    } finally {
      setLoading(false);
    }
  }, [orderBy]);

  useEffect(() => {
    loadContacts();
  }, [loadContacts]);

  function handleToggleOrderBy() {
    setOrderBy((prevState) => (prevState === 'asc' ? 'desc' : 'asc'));
  }

  function handleChangeSearchTerm(ev) {
    setSearchTerm(ev.target.value);
  }

  function handleTryAgain() {
    loadContacts();
  }

  return (
    <>
      <Container>
        <Loader isLoading={loading} />

        {contacts.length > 0 && (
          <InputSearchContainer>
            <input
              value={searchTerm}
              onChange={(ev) => handleChangeSearchTerm(ev)}
              type="text"
              placeholder="Procure pelo contato"
            />
          </InputSearchContainer>
        )}

        <Header justifyContent={
          hasError
            ? 'flex-end'
            : (contacts.length > 0
              ? 'space-between'
              : 'center')
        }
        >
          {(!hasError && contacts.length > 0) && (
            <strong>
              {filteredContacts.length > 0
                ? `${filteredContacts.length} ${filteredContacts.length > 1 ? ' contatos' : ' contato'
                }`
                : 'Nenhum contato!'}
            </strong>
          )}
          <Link to="/new">Novo Contato</Link>
        </Header>

        {hasError && (
          <ErrorContainer>
            <img src={sad} alt="Sad" />
            <div className="error_container">
              <strong>Ocorreu um erro ao obter os seus contatos!</strong>
              <Button onClick={handleTryAgain}>Tentar novamente</Button>
            </div>
          </ErrorContainer>
        )}

        {!hasError && (
          <>
            {contacts.length < 1 && (
              <NoContactsContainer>
                <img src={emptyBox} alt="Sem contatos" />
                <p>
                  Você ainda não tem nenhum contato cadastrado!
                  Clique no botão <strong>”Novo contato”</strong> à cima
                  para cadastrar o seu primeiro!
                </p>
              </NoContactsContainer>
            )}

            {filteredContacts.length > 0 && (
              <ListHeader orderBy={orderBy}>
                <button onClick={handleToggleOrderBy} type="button">
                  <span>Nome</span>
                  <img src={arrow} alt="Arrow" />
                </button>
              </ListHeader>
            )}

            {(contacts.length > 0 && filteredContacts.length < 1)
              && (
                <NotFoundFilteredContacts>
                  <img src={magnifierQuestion} alt="magnifierQuestion" />
                  <span>
                    Nenhum resultado foi encontrado para <b>{searchTerm}</b>.
                  </span>
                </NotFoundFilteredContacts>
              )}

            {filteredContacts.map((contact) => (
              <Card key={contact.id}>
                <div className="info">
                  <div className="contact-name">
                    <strong>{contact.name}</strong>
                    {contact.category_name && (
                      <small>{contact.category_name}</small>
                    )}
                  </div>
                  <span>{contact.email}</span>
                  <span>{contact.phone}</span>
                </div>
                <div className="actions">
                  <Link to={`/edit/${contact.id}`}>
                    <img src={edit} alt="edit" />
                  </Link>
                  <button type="button">
                    <img src={trash} alt="trash" />
                  </button>
                </div>
              </Card>
            ))}
          </>
        )}

      </Container>
    </>
  );
}
