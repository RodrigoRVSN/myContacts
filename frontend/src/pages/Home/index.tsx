import React from 'react';
import {
  Container,
} from './styles';

import { Loader } from '../../components/Loader';
import { Modal } from '../../components/Modal';
import { useHome } from './useHome';
import { InputSearch } from './components/InputSearch';
import { Header } from './components/Header';
import { ErrorStatus } from './components/ErrorStatus';
import { EmptyList } from './components/EmptyList';
import { SearchNotFound } from './components/SearchNotFound';
import { ContactsList } from './components/ContactsList';

export function Home() {
  const {
    isLoading,
    isDeleteModalVisible,
    contactBeingDeleted,
    handleToggleDeleteModal,
    handleConfirmDeleteContact,
    isLoadingDelete,
    contacts,
    searchTerm,
    handleChangeSearchTerm,
    hasError,
    filteredContacts,
    handleTryAgain,
    orderBy,
    handleToggleOrderBy,
    handleDeleteContact,
  } = useHome();

  const hasContacts = contacts.length > 0;
  const isListEmpty = !hasError && (!isLoading && !hasContacts);
  const isSearchEmpty = !hasError && hasContacts && filteredContacts.length < 1;

  return (
    <Container>
      <Loader isLoading={isLoading} />

      {hasContacts && (
        <InputSearch
          value={searchTerm}
          onChange={handleChangeSearchTerm}
        />
      )}

      <Header
        hasError={hasError}
        qtyContacts={contacts.length}
        qtyFilteredContacts={filteredContacts.length}
      />

      {hasError && <ErrorStatus onTryAgain={handleTryAgain} />}
      {isListEmpty && <EmptyList />}
      {isSearchEmpty && <SearchNotFound searchTerm={searchTerm} />}

      {hasContacts && (
      <>
        <ContactsList
          filteredContacts={filteredContacts}
          onDeleteContact={handleDeleteContact}
          orderBy={orderBy}
          onToggleOrderBy={handleToggleOrderBy}
        />

        <Modal
          danger
          isVisible={isDeleteModalVisible}
          title={`Tem certeza que deseja remover o contato "${contactBeingDeleted?.name}"?`}
          cancelLabel="Cancelar"
          confirmLabel="Confirmar"
          onCancel={handleToggleDeleteModal}
          onConfirm={handleConfirmDeleteContact}
          isLoading={isLoadingDelete}
        >
          <p>Esta ação não poderá ser desfeita!</p>
        </Modal>
      </>
      )}
    </Container>
  );
}
