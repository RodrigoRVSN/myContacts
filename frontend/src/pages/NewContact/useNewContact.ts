import { FormEventHandler, MutableRefObject, useRef } from 'react';
import ContactsService from '../../services/ContactsService';
import { DomainContact } from '../../types/IContact';
import { toast } from '../../utils/toast';

export const useNewContact = () => {
  const contactFormRef = useRef<{resetFields: () => void}>()

  async function handleSubmit(contact: DomainContact) {
    try {
      await ContactsService.createContact(contact);

      contactFormRef.current!.resetFields();

      toast({ type: 'success', text: 'Cadastro efetuado!', duration: 10000 });
    } catch (error) {
      toast({ type: 'danger', text: 'Houve um erro!' });
    }
  }

  return { contactFormRef, handleSubmit };
};

