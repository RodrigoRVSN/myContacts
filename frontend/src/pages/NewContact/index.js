import React from 'react';
import { PageHeader } from '../../components/PageHeader';
import { ContactForm } from '../../components/ContactForm';
import { useNewContact } from './useNewContact';

export function NewContact() {
  const { contactFormRef, handleSubmit } = useNewContact();

  return (
    <>
      <PageHeader title="Novo contato" />

      <ContactForm
        buttonLabel="Cadastrar"
        onSubmit={handleSubmit}
        ref={contactFormRef}
      />
    </>
  );
}
