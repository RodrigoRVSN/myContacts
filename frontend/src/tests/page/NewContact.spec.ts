import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import { NewContact } from '../../pages/NewContact';

describe('<NewContact />', () => {
  it('should render new contacts page', () => {
    const { container } = render(
      <Router>
        <NewContact />
      </Router>,
    );
    expect(container).toMatchSnapshot();
  });
});
