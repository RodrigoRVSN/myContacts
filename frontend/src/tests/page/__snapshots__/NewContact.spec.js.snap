// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewContact /> should render new contacts page 1`] = `
<div>
  <header
    class="sc-bdnxRM MRUKm"
  >
    <a
      href="/"
    >
      <img
        alt="arrow back"
        src="arrow.svg"
      />
      <span>
        Voltar
      </span>
    </a>
    <h1>
      Novo contato
    </h1>
  </header>
  <form
    class="sc-jSFjdj kLWweY"
    novalidate=""
  >
    <div
      class="sc-dlnjwi kuwTSj"
    >
      <input
        class="sc-hKFxyN hcprCk"
        placeholder="Nome *"
        value=""
      />
    </div>
    <div
      class="sc-dlnjwi kuwTSj"
    >
      <input
        class="sc-hKFxyN hcprCk"
        placeholder="E-mail"
        type="email"
        value=""
      />
    </div>
    <div
      class="sc-dlnjwi kuwTSj"
    >
      <input
        class="sc-hKFxyN hcprCk"
        maxlength="15"
        placeholder="Telefone"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-dlnjwi kuwTSj"
    >
      <select
        class="sc-eCApnc ljWfko"
      >
        <option
          value=""
        >
          Categoria
        </option>
        <option
          value="discord"
        >
          Discord
        </option>
        <option
          value="instagram"
        >
          Instagram
        </option>
      </select>
    </div>
    <div
      class="sc-gKAaRy hqCgCr"
    >
      <button
        class="sc-gtsrHT JklIe"
        disabled=""
        type="submit"
      >
        Cadastrar
      </button>
    </div>
  </form>
</div>
`;
